### PR TITLE
feat: compile static binary as release assets and docker image

### PR DESCRIPTION
**Pull request by zemyblue**
## Description

closes: #71 , #203 

### Changes
* add multi-platform docker image building, and push to docker hub.
* fix the bug that the `fnsad` version in docker image was not displayed.
* add the static fnsad binary compile function (just support `linux/amd64` and `linux/arm64`)
* change base docker image 

## Motivation and context

The release assets cannot execute without install dynamic library(libwasmvm). So it is very uncomfortable. 

## Screenshots (if appropriate):
The sample release assets results.
스크린샷 2023-06-20 오전 11 21 10

## Checklist:

- [x] I followed the [contributing guidelines](https://github.com/Finschia/finschia/blob/main/CONTRIBUTING.md) and [code of conduct](https://github.com/Finschia/finschia/blob/main/CODE_OF_CONDUCT.md).
- [x] I have added a relevant changelog to `CHANGELOG.md`
- [ ] I have added tests to cover my changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -1,53 +1,127 @@
...
         uses: docker/setup-buildx-action@v2
         with:
           endpoint: buildx
           config-inline: |
             [registry."docker.io"]
-      - name: login to the registry
+
+      - name: Login to the registry
         uses: docker/login-action@v2
         if: github.event_name != 'pull_request'
```
Comment (jaeseung-bae) [inline .github/workflows/docker.yml:38]: Since this github action doesn't get triggered by PR, can we remove it?
```suggestion

```

File: .github/workflows/docker.yml [hunk 2]
```diff
@@ -1,53 +1,127 @@
...
             type=semver,pattern={{version}}
-      - name: docker build
+
+      # Distroless Docker image (default)
+      - name: Build and push (distroless)
+        uses: docker/build-push-action@v4
+        with:
+          file: Dockerfile
+          context: .
+          push: ${{github.event_name != 'pull_request'}}
```
Comment (jaeseung-bae) [inline .github/workflows/docker.yml:81]: Since this github action doesn't get triggered by PR, can we change this?
```suggestion
          push: true
```

File: README.md [hunk 3]
```diff
@@ -18,18 +18,18 @@ This repository hosts `Finschia`. This repository is forked from [gaia](https://
...
-make build-docker WITH_CLEVELDB=yes GITHUB_TOKEN=${YOUR_GITHUB_TOKEN}  # build docker image with cleveldb
+make docker-build WITH_CLEVELDB=yes GITHUB_TOKEN=${YOUR_GITHUB_TOKEN}  # build docker image with cleveldb
 ```
 
 _Note1_
 
 If you are using M1 mac, you need to specify build args like this:
 ```
-make build-docker ARCH=aarch64
+make docker build ARCH=aarch64
```
Comment (jaeseung-bae) [inline README.md:32]: May need chage?
```suggestion
make docker-build ARCH=aarch64
```

**Review by jaeseung-bae — APPROVED** [review]

File: .github/workflows/docker.yml [hunk 4]
```diff
@@ -1,53 +1,126 @@
...
-      - main
   push:
     tags:
       - "v[0-9]+.[0-9]+.[0-9]+" # Push events to matching v*, i.e. v1.0, v20.15.10
       - "v[0-9]+.[0-9]+.[0-9]+-*" # Push events to matching v*, i.e. v1.0-rc1, v20.15.10-alpha1
 
 env:
   registry: docker.io
-  repository: finschia/finschianode
+  DOCKER_REPOSITORY: finschia/finschianode
```
Comment (0Tech) [inline .github/workflows/docker.yml:12]: Why should we provide multiple docker images for the same binary?
Can you elaborate the background?

**Review by 0Tech — COMMENTED** [review]
We get similar binaries by several logics, by Makefile or by Dockerfile. This pr adds the way even more. We should consider tidy up the logics. We can reuse some logics in the other logics.
Also, we should concentrate on the main targets. The others just increases our burden on maintenance & QA.

File: .github/workflows/docker.yml (hunk 4 above)
Comment (zemyblue) [reply .github/workflows/docker.yml:12]: Because the optimal version for each system  runs faster. 
For example, in my case, I use M1 mac. So when I use `linux/arm64`, the docker image more fast then `linux/amd64`.

Comment (0Tech) [reply .github/workflows/docker.yml:12]: The point is not about architectures, but about the base image. Architecture support is important feature.
You've added 3 base images. But I think one image would be sufficient.

Comment (zemyblue) [reply .github/workflows/docker.yml:12]: Our current system and test are fitted with `alpine` base image. But as you said, `alpine` has an elementf of insecurity, and I also think it's good to use `debian11` when we use just `fnsad` cli without any compile in any platform, even windows. So I added `debian11` and `debian11-noroot` base image. Do you think it is no need?

**Comment by zemyblue** [comment]
> We get similar binaries by several logics, by Makefile or by Dockerfile. This pr adds the way even more. We should consider tidy up the logics. We can reuse some logics in the other logics.
> Also, we should concentrate on the main targets. The others just increases our burden on maintenance & QA.

Yes, there are meny build option and it is a little complex. I also think It need to arrange it. I'll add as a issue it. And I'll try to fix it as another pr.

File: .github/workflows/docker.yml (hunk 4 above)
Comment (zemyblue) [reply .github/workflows/docker.yml:12]: But I also think other base image is no need. Because alpine can do it and replace it. 
I'll leave `alpine` and delete the rest. Thank you.

Comment (0Tech) [reply .github/workflows/docker.yml:12]: > But as you said, `alpine` has an elementf of insecurity,

I've not said about any security concerns, please check the corresponding docker hub (https://hub.docker.com/_/golang). So I think we should have a discussion about it.

> it's good to use `debian11` when we use just `fnsad` cli without any compile in any platform, even windows.

Technically, any "good" base image would allow running the binary. One of the reasons of using container is encapsulating the running environment, so it's OK to use shared library on it. However, combining with the cosmovisor, it would be better to make it static. Then, practically, any linux kernel based container would be sufficient.

TL;DR
1. It doesn't matter whether the executable is static or not, in containers.
2. However, with the cosmovisor, we should provide the static executables.

In my opinion, we should build the static executable, and our containers should contain that executable (for maintenance reason).

> So I added `debian11` and `debian11-noroot` base image. Do you think it is no need?

Can you elaborate the differences of the both images?

Comment (zemyblue) [reply .github/workflows/docker.yml:12]: `Distroless` debian11 image does not contain package managers and shell script, and we can just contain `fnsad` and runtime dependencies. So it is small size and I think we can generate more secure image in terms of security.
So I add this base image in our docker image. 

https://github.com/GoogleContainerTools/distroless

Comment (0Tech) [reply .github/workflows/docker.yml:12]: OK. I think we can stick to alpine for now. Using golang:alpine seems a little bit tricky, but also OK for now because it's working. We may address it when the related issue arises.

**Review by 0Tech — DISMISSED** [review]

**Review by jaeseung-bae — APPROVED** [review]

**Review by jaeseung-bae — APPROVED** [review]

**Review by 0Tech — APPROVED** [review]
